### PR TITLE
Add support for sqlite3_unlock_notify

### PR DESCRIPTION
**Pull request by typeless**
For https://github.com/go-gitea/gitea/issues/2040

Added new build tag `sqlite_unlock_notify`.

```
name      old req/s   new req/s   delta            
Exec       219k ±20%   217k ±18%     ~     (p=0.690 n=5+5)                                            
Query     75.8k ±12%  67.9k ±13%     ~     (p=0.095 n=5+5)                                            
Params    74.1k ± 3%  68.5k ± 6%   -7.63%  (p=0.008 n=5+5)                                            
Stmt       119k ± 1%   108k ± 5%   -9.56%  (p=0.008 n=5+5)                                            
Rows      2.21k ± 1%  2.08k ± 1%   -5.61%  (p=0.016 n=5+4)                                            
StmtRows  2.30k ± 2%  2.05k ±11%  -10.78%  (p=0.008 n=5+5)  
```

### CODE REVIEW

**Comment by mattn** [comment]
kick travis

**Comment by mattn** [comment]
Could you please rebase ?

**Comment by typeless** [comment]
I'd like to add a test for what the PR tries to fix but cannot deliver it in the first few attempts. I'll try later when I have more free time.

**Comment by GJRTimmer** [comment]
@typeless Could you rebase

**Comment by typeless** [comment]
@GJRTimmer Done

**Comment by mattn** [comment]
I wonder this can be tested.

File: sqlite3.go [hunk 1]
```diff
@@ -81,8 +82,25 @@ _sqlite3_exec(sqlite3* db, const char* pcmd, long long* rowid, long long* change
 static int
 _sqlite3_step(sqlite3_stmt* stmt, long long* rowid, long long* changes)
 {
-  int rv = sqlite3_step(stmt);
+  extern int unlock_notify_wait(sqlite3 *db);
+  int rv;
   sqlite3* db = sqlite3_db_handle(stmt);
+
+  for (;;) {
```
Comment (GJRTimmer) [inline sqlite3.go]: Shouldn't this be
```go
for {
```

**Comment by GJRTimmer** [comment]
@mattn I think you are right we need some code to test it, maybe even an example.

@typeless Instead of currently trying to write a test for it, You did probably design this with some code to see if it worked. Could you add an example for this to the `_example` folder ?

File: sqlite3.go (hunk 1 above)
Comment (typeless) [reply sqlite3.go]: I don't have access to the dev environment right now but this `for` looks like C code.

**Comment by bvinc** [comment]
I notice that in the step function, you check the extended error code to make sure it's `SQLITE_LOCKED_SHAREDCACHE`, but you don't do this in the prepare code.  Why not?

**Comment by typeless** [comment]
@bvinc It was modified from the example code:  https://sqlite.org/unlock_notify.html

The explanation of the `SQLITE_LOCKED_SHAREDCACHE` check is described in the last few paragraphs.
It seems to me that `sqlite3_blocking_prepare_v2` would not result in the same scenario as `qlite3_blocking_step` does.

**Comment by typeless** [comment]
By the way, I have found a test which can reproduce the "database is locked" error but it turned out that the PR does not fix the problem. I'll try to figure it out later and see if this PR has any hope to be helpful.

**Comment by bvinc** [comment]
I found your test case in a closed PR.  I assume this is the test case that you're talking about.  I believe that there is some confusion again between `SQLITE_BUSY` and `SQLITE_LOCKED`.

Without using shared cache mode (which applies to your test), when two transactions conflict, the one that fails uses the busy handler, which waits a small number of random milliseconds, and then tries again.  Once this fails for 5 seconds (which is the default busy timeout for this driver), then it fails with `SQLITE_BUSY`, which says `database is locked`.  What's happening with your test case is that one statement is spinning so fast that it's completely starving out the other statement for a full 5 seconds.  This is completely normal and just how SQLite works.  You can see that if you increase the busy timeout by adding "?_busy_timeout=60000" that it won't give an error anymore.

But *this* PR is about something completely different.  This PR is about stopping people from getting `SQLITE_LOCKED` when using shared cache mode (which your test case did not use).

**Comment by typeless** [comment]
@bvinc You're correct I misunderstood the two different errors. I need to try it again when I come back to this.

**Comment by typeless** [comment]
@mattn @GJRTimmer @bvinc  The test has been added. PTAL.

**Comment by typeless** [comment]
Hmmm, I have to make it compatible with `USE_LIBSQLUTE3` .

**Comment by typeless** [comment]
Found a bug while trying to integrate the change into Gitea. I am looking into it.

**Comment by mattn** [comment]
Thank you. Sounds good to me. One thing. please add `_` for the functions always to be not use same namespace `sqlite_`.

**Comment by typeless** [comment]
@mattn Done. Is that what you mean?

**Comment by mattn** [comment]
Yes! Thank you.

**Pull request merged**